src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,688 @@
+        {
+            var actionsBuilder = new ArrayBuilder<CodeAction>();
+            var (isParameterized, hasOptionalParameter) = ExpressionWithinParameterizedMethod(semanticDocument, expression, cancellationToken);
+            if (isParameterized)
+            {
+                actionsBuilder.Add(new MyCodeAction(CreateDisplayText(allOccurrences: false, trampoline: false, overload: false),
+                    c => IntroduceParameterAsync(semanticDocument, expression, false, false, false, cancellationToken)));
+                actionsBuilder.Add(new MyCodeAction(CreateDisplayText(allOccurrences: true, trampoline: false, overload: false),
+                    c => IntroduceParameterAsync(semanticDocument, expression, true, false, false, cancellationToken)));
+
+                if (!hasOptionalParameter)
+                {
+                    actionsBuilder.Add(new MyCodeAction(CreateDisplayText(allOccurrences: false, trampoline: true, overload: false),
+                        c => IntroduceParameterAsync(semanticDocument, expression, false, true, false, cancellationToken)));
+                    actionsBuilder.Add(new MyCodeAction(CreateDisplayText(allOccurrences: true, trampoline: true, overload: false),
+                        c => IntroduceParameterAsync(semanticDocument, expression, true, true, false, cancellationToken)));
+                    actionsBuilder.Add(new MyCodeAction(CreateDisplayText(allOccurrences: false, trampoline: false, overload: true),
+                        c => IntroduceParameterAsync(semanticDocument, expression, false, false, true, cancellationToken)));
+                    actionsBuilder.Add(new MyCodeAction(CreateDisplayText(allOccurrences: true, trampoline: false, overload: true),
+                        c => IntroduceParameterAsync(semanticDocument, expression, true, false, true, cancellationToken)));
+                }
+            }
+
+            return actionsBuilder.ToImmutable();
+        }
+
+        private static string CreateDisplayText(bool allOccurrences, bool trampoline, bool overload)
+                => (allOccurrences, trampoline, overload) switch
+                {
+                    (true, true, false) => FeaturesResources.Extract_method_to_invoke_at_all_callsites_for_all_occurrences_of_0,
+                    (true, false, false) => FeaturesResources.Introduce_parameter_for_all_occurrences_of_0,
+                    (true, false, true) => FeaturesResources.Introduce_overload_with_new_parameter_for_all_occurrences_of_0,
+                    (false, true, false) => FeaturesResources.Extract_method_to_invoke_at_all_callsites_for_0,
+                    (false, false, true) => FeaturesResources.Introduce_overload_with_new_parameter_for_0,
+                    (false, false, false) => FeaturesResources.Introduce_parameter_for_0,
+                    _ => throw new System.NotImplementedException()
+                };
```

Comment:
this function seems superfluous.  it is always called with constants, and each constant maps to a constant result.  so you can just inline the results at the callpoints.

Reply:
removed